core/os/os.cpp

```diff
@@ -328,17 +328,16 @@ void OS::yield() {
 
 void OS::ensure_user_data_dir() {
 	String dd = get_user_data_dir();
-	DirAccess *da = DirAccess::open(dd);
-	if (da) {
-		memdelete(da);
-		return;
+	{
+		DirAccessRef da = DirAccess::open(dd);
+		if (da) {
+			return;
+		}
 	}
```

Comment:
Note: Initially this wasn't scoped and `da` was reused to assign the new filesystem access. That crashed.
Sounds like a potential bug in `DirAccessRef` to me, but until then it's a lesson not to reuse `DirAccessRef` variables.

Reply:
I'm getting this error with this:
```
platform/iphone/export/export_plugin.cpp:792:62: error: use of deleted function 'constexpr DirAccessRef::DirAccessRef(const DirAccessRef&)'
  792 |   Error err = _walk_dir_recursive(p_da, p_handler, p_userdata);
      |                                                              ^
In file included from ./editor/editor_export.h:34,
                 from platform/iphone/export/export_plugin.h:43,
                 from platform/iphone/export/export_plugin.cpp:31:
./core/io/dir_access.h:130:8: note: 'constexpr DirAccessRef::DirAccessRef(const DirAccessRef&)' is implicitly declared as deleted because 'DirAccessRef' declares a move constructor or move assignment operator
  130 | struct DirAccessRef {
      |        ^~~~~~~~~~~~
platform/iphone/export/export_plugin.cpp:768:65: note:   initializing argument 1 of 'static Error EditorExportPlatformIOS::_walk_dir_recursive(DirAccessRef, EditorExportPlatformIOS::FileHandler, void*)'
  768 | Error EditorExportPlatformIOS::_walk_dir_recursive(DirAccessRef p_da, FileHandler p_handler, void *p_userdata) {
      |                                                    ~~~~~~~~~~~~~^~~~
```

Reply:
I guess it's trying to copy it anyway, probably passing it to the functions should be by reference, `DirAccessRef &p_da`.

Reply:
For now I decided to keep it simple and reverted `EditorExportPlatformIOS::_walk_dir_recursive()` and `EditorFileSystem::_scan_new_dir()` to use `DirAccess *`.

Reply:
Are there any reasons to use custom `DirAccessRef` at all, maybe it's better to make `DirAccess` ref. counted, and use `Ref<DirAccess>`. In this case, duplicate `Directory` core bind can be removed as well.

Reply:
And, I guess, the same question for the `core_bind` in general, is it necessary at all?

Reply:
After testing I decided to remove the move semantics as I think they might actually lead to leaking `DirAccess *`, which is what `DirAccessRef` tries to prevent. `SWAP` is likely not enough, the lvalue's `f` should be deleted too.

I tested with this:
```diff
diff --git a/core/os/os.cpp b/core/os/os.cpp
index 117f719b04..84cb65788c 100644
--- a/core/os/os.cpp
+++ b/core/os/os.cpp
@@ -329,11 +329,18 @@ void OS::yield() {
 void OS::ensure_user_data_dir() {
 	String dd = get_user_data_dir();
 	DirAccessRef da = DirAccess::open(dd);
+	DirAccess *f = da.f;
 	if (da) {
-		return;
+		//return;
 	}
 
+	if (f) {
+		printf("1: %ld\n", (int64_t)f);
+	}
 	da = DirAccess::create(DirAccess::ACCESS_FILESYSTEM);
+	if (f) {
+		printf("2: %ld\n", (int64_t)f);
+	}
 	Error err = da->make_dir_recursive(dd);
 	ERR_FAIL_COND_MSG(err != OK, "Error attempting to create data dir: " + dd + ".");
 }
```

and got:
```
1: 186029184
2: 186029184
```

While I'd want `f` to be `nullptr` after it has been replaced.

Reply:
Everything seems to be correct, `f` is copy of pointer, why would it change? But after `da = ...` it should point to the invalid memory.

Reply:
> Are there any reasons to use custom `DirAccessRef` at all, maybe it's better to make `DirAccess` ref. counted, and use `Ref<DirAccess>`. In this case, duplicate `Directory` core bind can be removed as well.

That could be nice yes. @reduz hinted at something similar on #core.
That can likely be done in a follow-up PR with easy replace of the current `DirAccessRef`.

> Everything seems to be correct, `f` is copy of pointer, why would it change? But after `da = ...` it should point to the invalid memory.

Ah right, my bad. But as of the current state of this PR I don't need it anymore, and if we're going to make it ref. counted I guess it's not necessary to implement it now.

Reply:
> And, I guess, the same question for the `core_bind` in general, is it necessary at all?

Do you mean `core_bind::Directory`? That's the one exposed to the scripting API. It could maybe be unified I guess those the `Directory` API is a bit higher level.

Reply:
I mean everything in the `core_bind`. If `DirAccess` will be changed to RefCounted, it can be exposed to the scripting API directly, there's no need to have `Directory`. The same is true for the rest of `core_bind` classes, it seems to be an excessive extra level of abstraction.